### PR TITLE
[travis] Added jslint check and fixed javascript

### PR DESCRIPTION
**Pull request by HizkiFW**
The js files had to be modified to prevent jslint from giving up early.

### CODE REVIEW

File: django_netjsonconfig/static/django-netjsonconfig/js/preview.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
+/*global closePreview */
```
Comment (nemesisdesign) [inline django_netjsonconfig/static/django-netjsonconfig/js/preview.js]: add this to `predef` in `.jslintrc` and get rid of this addition

File: django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js [hunk 2]
```diff
@@ -1,3 +1,5 @@
+/*global objectIsEqual */
+
```
Comment (nemesisdesign) [inline django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js]: same here

File: django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js [hunk 3]
```diff
@@ -27,28 +29,29 @@
                     try{
                         object[name] = JSON.parse(value);
                     }
-                    catch(e){}
+                    catch(ignore){}
```
Comment (nemesisdesign) [inline django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js]: you may try as well `catch(){}`

File: django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js [hunk 4]
```diff
@@ -27,28 +29,29 @@
...
             initialField = $('#initial-id_' + name);
             initialValue = initialField.length ? initialField.val() : django._njc_initial_values[name];
             // fix checkbox value inconsistency
-            if (initialValue == 'True') { initialValue = true }
-            else if (initialValue == 'False') { initialValue = false }
-            if (name == 'config') { initialValue = JSON.parse(initialValue) }
+            if (initialValue == 'True') { initialValue = true; }
+            else if (initialValue == 'False') { initialValue = false; }
+            if (name == 'config') { initialValue = JSON.parse(initialValue); }
             // currentValue = current_values[name]
```
Comment (nemesisdesign) [inline django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js]: mm `currentValue` looks like dead forgotten code, you may remove all occurrences of it if it's true

File: django_netjsonconfig/static/django-netjsonconfig/js/widget.js [hunk 5]
```diff
@@ -1,3 +1,5 @@
+/*global editor, advancedJSONEditor, JSONEditor */
+
```
Comment (nemesisdesign) [inline django_netjsonconfig/static/django-netjsonconfig/js/widget.js]: same here

File: django_netjsonconfig/static/django-netjsonconfig/js/widget.js [hunk 6]
```diff
@@ -505,9 +510,8 @@ JSONEditor.defaults.themes.django = JSONEditor.AbstractTheme.extend({
             if (matchKey) {
                 if (elem[matchKey](selector)) {
                     return elem;
-                } else {
-                    elem = elem.parentNode;
```
Comment (nemesisdesign) [inline django_netjsonconfig/static/django-netjsonconfig/js/widget.js]: I don't understand, wouldn't this change also the behaviour of the code?

File: django_netjsonconfig/static/django-netjsonconfig/js/widget.js [hunk 7]
```diff
@@ -226,16 +228,16 @@
...
 var matchKey = (function () {
     var elem = document.documentElement;
-    if (elem.matches) return 'matches';
-    else if (elem.webkitMatchesSelector) return 'webkitMatchesSelector';
-    else if (elem.mozMatchesSelector) return 'mozMatchesSelector';
-    else if (elem.msMatchesSelector) return 'msMatchesSelector';
-    else if (elem.oMatchesSelector) return 'oMatchesSelector';
-})();
+    if (elem.matches) { return 'matches'; }
+    if (elem.webkitMatchesSelector) { return 'webkitMatchesSelector'; }
```
Comment (nemesisdesign) [inline django_netjsonconfig/static/django-netjsonconfig/js/widget.js]: I don't think changing all the `else if` to `if` is a good idea

File: .travis.yml [hunk 8]
```diff
@@ -24,13 +24,15 @@ before_install:
...
   - ./runisort
+  - npm install -g jslint
 
 install:
   - pip install $DJANGO
   - pip install https://github.com/openwisp/django-x509/tarball/master
   - python setup.py -q develop
 
 script:
+  - jslint django_netjsonconfig/static/django-netjsonconfig/js/*.js
```
Comment (nemesisdesign) [inline .travis.yml]: move the jslint check at the end of `before_install`

File: .jslintrc [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+	"predef": [
+		"document",
+		"navigator",
+		"console",
+		"window",
+		"django",
+		"gettext",
+		"windowname_to_id",
+		"jQuery",
+		"WebSocket",
```
Comment (nemesisdesign) [inline .jslintrc]: you can remove this

File: .jslintrc [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+		"$",
+		"dismissRelatedLookupPopup",
+		"dismissAddAnotherPopup",
+		"FileReader",
+		"Image",
+		"L",
```
Comment (nemesisdesign) [inline .jslintrc]: you can remove the previous 5 lines

File: .jslintrc [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+	"asi": true,
+	"sloppy": true,
+	"vars": true,
+    "forin": true,
+    "nomen": true,
+    "continue": true,
+    "plusplus": true,
+    "newcap": true,
+    "indent": 4,
+	"maxerr": 10000
```
Comment (nemesisdesign) [inline .jslintrc]: notice how the indentation is inconsistent, fix it please

File: django_netjsonconfig/static/django-netjsonconfig/js/widget.js (hunk 7 above)
Comment (HizkiFW) [reply django_netjsonconfig/static/django-netjsonconfig/js/widget.js]: In this case, when the `if` matches, it returns so the rest of the `if`s are never reached. It's the same as using an `else if`, but without jslint complaining 😛

File: django_netjsonconfig/static/django-netjsonconfig/js/widget.js (hunk 6 above)
Comment (HizkiFW) [reply django_netjsonconfig/static/django-netjsonconfig/js/widget.js]: Returning would mean the `elem = elem.parentNode;` part is never reached. The same happens with else. If the condition is matched, run `return elem;`, if not then run `elem = elem.parentNode;` and the rest. With the `return` statement, if the condition is met, the function returns and the rest of the code is never reached. If it does not, it'll continue on executing the rest of the code, including the `elem = elem.parentNode;` part.

There's a good explanation on StackOverflow: https://stackoverflow.com/a/3261899

File: django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js (hunk 3 above)
Comment (HizkiFW) [reply django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js]: jslint dies there if the `ignore` is deleted.
```
 #1 Expected an identifier and instead saw ')'.
    catch(){} // Line 30, Pos 27
 #2 Stopping. (28% scanned).
     // Line 30, Pos 27
```

File: .jslintrc (hunk 9 above)
Comment (HizkiFW) [reply .jslintrc]: `gettext` is used in `preview.js` (line 50 and 71) and `unsaved_changes.js` (line 43), so I've left it there.

File: django_netjsonconfig/static/django-netjsonconfig/js/widget.js (hunk 7 above)
Comment (nemesisdesign) [reply django_netjsonconfig/static/django-netjsonconfig/js/widget.js]: correct!

File: django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js (hunk 3 above)
Comment (nemesisdesign) [reply django_netjsonconfig/static/django-netjsonconfig/js/unsaved_changes.js]: ok

File: django_netjsonconfig/static/django-netjsonconfig/js/widget.js (hunk 6 above)
Comment (nemesisdesign) [reply django_netjsonconfig/static/django-netjsonconfig/js/widget.js]: correct, sorry I was going too fast

**Pull request merged**